scipy/linalg/src/_batched_linalg.h

```diff
@@ -0,0 +1,305 @@
+#include "_lapack_trampolines.h"
+#include "_npymath.h"
+
+
+using namespace _numpymath;
+
+
+// parroted from sqrtm
+// XXX can probably be replaced by ?copy from BLAS
+template<typename T>
+inline void
+swap_cf(T* src, T* dst, const Py_ssize_t r, const Py_ssize_t c, const Py_ssize_t n)
+{
+    Py_ssize_t i, j, ith_row, r2, c2;
+    T *bb = dst;
+    T *aa = src;
+    if (r < 16) {
+        for (j = 0; j < c; j++)
+        {
+            ith_row = 0;
+            for (i = 0; i < r; i++) {
+                bb[ith_row] = aa[i];
+                ith_row += n;
+            }
+            aa += n;
+            bb += 1;
+        }
+    } else {
+        // If tall
+        if (r > c)
+        {
+            r2 = r/2;
+            swap_cf(src, dst, r2, c, n);
+            swap_cf(src + r2, dst+(r2)*n, r-r2, c, n);
+        } else {  // Nope
+            c2 = c/2;
+            swap_cf(src, dst, r, c2, n);
+            swap_cf(src+(c2)*n, dst+c2, r, c-c2, n);
+        }
+    }
+}
+
+
+/*
+ * Visit (m, n)-shaped slices of a ndim>=3 array in C order.
+ * Copy each slice into a buffer, make the buffer F-ordered
+ * for LAPACK.
+ */
+struct iter_data_t
+{
+    // grab useful data from the array to iterate
+    npy_intp ndim;
+    npy_intp *shape;
+    npy_intp *strides;
+    void *data_ptr;
+
+    npy_intp outer_size;  // math.prod(a.shape[:-2])
+    npy_intp m, n;        // core dimensions
+
+    iter_data_t(PyArrayObject *arr) {
+        ndim = PyArray_NDIM(arr);
+        shape = PyArray_SHAPE(arr);
+        strides = PyArray_STRIDES(arr);
+        data_ptr = PyArray_DATA(arr);
+
+        m = shape[ndim - 2];
+        n = shape[ndim - 1];
+
+        /* outer_size: the number of slices in the batch, math.prod(a.shape[:-2] */
+        outer_size = 1;
+        if(ndim > 2) {
+            for(int i=0; i < ndim - 2; i++) {
+                outer_size *= shape[i];
+            }
+        }
+    }
+
+    /*
+     * Copy the slice number `idx` into the `buffer`.
+     */
+    template<typename T>
+    void copy_slice(const npy_intp idx, T *buffer) {
+
+        assert((0 <= idx) && (idx < outer_size));
+
+        /* parroted from sqrtm*/
+        npy_intp offset = 0;
+        npy_intp temp_idx = idx;
+        for (int i = ndim - 3; i >= 0; i--) {
+            offset += (temp_idx % shape[i]) * strides[i];
+            temp_idx /= shape[i];
+        }
+
+        T *slice_ptr = (T *)(data_ptr) + offset/sizeof(T);
+
+        // copy the current n-x-n slice into the temp storage *in the Fortran order*
+        for(npy_intp i=0; i < m; i++) {
+            for(npy_intp j=0; j < n; j++) {
+                buffer[i + j*m] = *(slice_ptr + (i*strides[ndim - 2]/sizeof(T)) + (j*strides[ndim - 1]/sizeof(T)));
+            }
+        }
+    }
+};
+
+
+/***********************************
+ *   Inversion via GETRF / GETRI
+ ***********************************/
+
+/*
+ * Invert a 2D slice:
+ *   - Input slice is in `getrf_data.a`.
+ *   - The result is in `getri_data.a`.
```

Comment:
This is a class instance here. I really don't see the complication helping us here a class with a single method is pretty much a function. Still missing the call to `gecon` though.